src/machine-c-api.h

```diff
@@ -451,1508 +159,435 @@ typedef struct { // NOLINT(modernize-use-using)
-CM_API int cm_write_mcause(cm_machine *m, uint64_t val, char **err_msg);
+CM_API int cm_write_uarch_cycle(cm_machine *m, uint64_t val);
 
-/// \brief Reads the value of the mtval register.
+/// \brief Gets the value of the microarchitecture halt flag.
 /// \param m Pointer to valid machine instance
-/// \param val Receives value of the register.
-/// \param err_msg Receives the error message if function execution fails
-/// or NULL in case of successful function execution. In case of failure error_msg
-/// must be deleted by the function caller using cm_delete_cstring.
-/// err_msg can be NULL, meaning the error message won't be received.
+/// \param val Receives value of the halt flag.
 /// \returns 0 for success, non zero code for error
-CM_API int cm_read_mtval(const cm_machine *m, uint64_t *val, char **err_msg);
+CM_API int cm_read_uarch_halt_flag(const cm_machine *m, bool *val);
 
-/// \brief Writes the value of the mtval register.
+/// \brief Sets the value of the microarchitecture halt flag.
 /// \param m Pointer to valid machine instance
-/// \param val New register value.
-/// \param err_msg Receives the error message if function execution fails
-/// or NULL in case of successful function execution. In case of failure error_msg
-/// must be deleted by the function caller using cm_delete_cstring.
-/// err_msg can be NULL, meaning the error message won't be received.
 /// \returns 0 for success, non zero code for error
-CM_API int cm_write_mtval(cm_machine *m, uint64_t val, char **err_msg);
+CM_API int cm_set_uarch_halt_flag(cm_machine *m);
 
-/// \brief Reads the value of the misa register.
-/// \param m Pointer to valid machine instance
-/// \param val Receives value of the register.
-/// \param err_msg Receives the error message if function execution fails
-/// or NULL in case of successful function execution. In case of failure error_msg
-/// must be deleted by the function caller using cm_delete_cstring.
-/// err_msg can be NULL, meaning the error message won't be received.
-/// \returns 0 for success, non zero code for error
-CM_API int cm_read_misa(const cm_machine *m, uint64_t *val, char **err_msg);
+////////////////////////////////////////////////////////////////////////////////
+// Getting addresses
 
-/// \brief Writes the value of the misa register.
-/// \param m Pointer to valid machine instance
-/// \param val New register value.
-/// \param err_msg Receives the error message if function execution fails
-/// or NULL in case of successful function execution. In case of failure error_msg
-/// must be deleted by the function caller using cm_delete_cstring.
-/// err_msg can be NULL, meaning the error message won't be received.
+/// \brief Returns count of memory ranges in the machine.
+/// \param count Receives the memory range count.
 /// \returns 0 for success, non zero code for error
-CM_API int cm_write_misa(cm_machine *m, uint64_t val, char **err_msg);
+CM_API int cm_get_memory_range_count(cm_machine *m, int *count);
```

Comment:
Cleaver trick.

But one goal here is to not ask users who is using the API or creating a binding to perform allocations on their own, some might not know how to use `malloc/free` well, in fact some people using the C API of the emulator confused before who has the ownership of these struct pointers, either they forget to free thinking the ownership moved, or caused double free.

If we do the suggested cleaver way, it would also be the only struct needed, which would increase the logic of the binding generators just because of one struct, we might as well get rid of it. Ideally I was thinking of an API like OpenGL, who gets away without any struct. The proposed memory range API I did has two functions but I think it is more fail proof of users mistakes and more easy on the binding generator work despite adding a new function.  Also people binding the emulator C API complained that the binding work of structs are boilerplate.

Reply:
Just return a JSON with all the fields there.

Reply:
I was trying to avoid returning a JSON for this specific case, otherwise people will really need to include a JSON parser in case they want to iterate over a machine PMAs. The same problem happens for jsonrpc semantic version return.

The proposed API changes a C user could use most of the API without needing to include a JSON parser library.

Thinking about people using the API with just C, asking them to depend on a JSON parser is kinda annoying in my view, but well, maybe they will not need to iterate over PMAs anyway.

In my view JSON strings only make sense for configs, access logs and proofs, not more than that.

Reply:
My worry is that the API won't be ergonomic for "lower" level languages, just the high level ones.
But ok, you comment makes sense since you managed to remove the other structs.

Reply:
> My worry is that the API won't be ergonomic for "lower" level languages
Agreed, but calling these 2 functions still easy to use in low level languages. However returning JSON it's less easy because you have to pull a JSON parser.

Reply:
Either we have the structs, or we don't. If we don't have the structs, you need a JSON parser. I don't see the distinction here. If we are going to do it, let's go all the way. Or are you saying that iterating over the memory ranges is something that is more common than getting the default config or getting a proof or getting a step log or some such?

Reply:
I am assuming every bind will be to a language that has JSON support. The only language that will suffer is C, right? Do we have examples other than C? Is anybody really going to use C?

Reply:
> I am assuming every bind will be to a language that has JSON support. The only language that will suffer is C, right? Do we have examples other than C? Is anybody really going to use C?

Even C has it, but that is not the point.
Go and Rust handle C structs great, Dynamic languages maybe not so well (However Node could bind the C++ API directly), Lua is already contemplated, which leaves out Python.

I believe that manual memory management is more of an issue since none of the above has it. At least from what I remember on the subject.

This opinion may be a little late on this discussion, but my worry is that we are optimizing for the wrong thing.

Reply:
> Either we have the structs, or we don't. If we don't have the structs, you need a JSON parser. I don't see the distinction here. If we are going to do it, let's go all the way. Or are you saying that iterating over the memory ranges is something that is more common than getting the default config or getting a proof or getting a step log or some such?

I think it is, if I use the C API in C program, I might not care about getting initial configs, or proofs, or access log of a stored machine. But I might care checking what memory ranges the loaded machine has.

Reply:
This is something you will do once per machine you load. It’s not at issue regarding performance. Nobody will use the C API directly.

Reply:
Ok, I will change this to be also a JSON.

Reply:
Removed the two methods and introduced `cm_get_memory_ranges` returning a JSON string in latest commit.

Reply:
Sure!